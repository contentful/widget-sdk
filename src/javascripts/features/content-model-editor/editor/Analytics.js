import * as Analytics from 'analytics/Analytics';
import { get } from 'lodash';

export const trackAddedField = (contentType, field) => {
  Analytics.track('content_modelling:field_added', {
    contentTypeId: contentType.sys.id,
    contentTypeName: contentType.name,
    fieldId: field.id,
    fieldName: field.name,
    fieldType: field.type,
    fieldItemType: get(field, 'items.type') || null,
    fieldLocalized: field.localized,
    fieldRequired: field.required,
  });
};

export const trackEnforcedButtonClick = (err) => {
  // If we get reason(s), that means an enforcement is present
  const reason = get(err, 'body.details.reasons', null);

  Analytics.track('entity_button:click', {
    entityType: 'contentType',
    enforced: Boolean(reason),
    reason,
  });
};
