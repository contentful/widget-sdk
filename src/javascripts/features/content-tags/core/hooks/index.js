export { useCreateTag } from 'features/content-tags/core/hooks/useCreateTag';
export { useUpdateTag } from 'features/content-tags/core/hooks/useUpdateTag';
export { useDeleteTag } from 'features/content-tags/core/hooks/useDeleteTag';
export { useReadTags } from 'features/content-tags/core/hooks/useReadTags';
export { useF36Modal } from 'features/content-tags/core/hooks/useF36Modal';
export { useScrollToTop } from 'features/content-tags/core/hooks/useScrollToTop';
export { useTagsRepo } from 'features/content-tags/core/hooks/useTagsRepo';
export { useCurrentSpaceFeature } from 'features/content-tags/core/hooks/useCurrentSpaceFeature';
export { useTagsFeatureEnabled } from 'features/content-tags/core/hooks/useTagsFeatureEnabled';
export { useSpaceContext } from 'features/content-tags/core/hooks/useSpaceContext';
export { useIsAdmin } from 'features/content-tags/core/hooks/useIsAdmin';
export { useIsInitialLoadingOfTags } from 'features/content-tags/core/hooks/useIsInitialLoadingOfTags';
