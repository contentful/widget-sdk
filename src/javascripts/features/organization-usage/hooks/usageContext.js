import React from 'react';
import { mapResponseToState } from '../services/UsageService';
import { periodToDates } from '../utils/periodToDates';
import { sum, get } from 'lodash';
import PropTypes from 'prop-types';

export const colours = ['#2E75D4', '#0EB87F', '#EA9005', '#8C53C2', '#CC3C52'];

const initialState = {
  isLoading: true,
  error: null,
  isAssetBandwidthTab: false,
  selectedMainTab: 'apiRequest',
  selectedPeriodIndex: 0,
  selectedSpacesTab: 'cma',
};

export const UsageStateContext = React.createContext();
export const UsageDispatchContext = React.createContext();

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_ORG_DATA': {
      const { periods } = action.value;
      const periodDates = periodToDates(periods[state.selectedPeriodIndex]);
      return {
        ...state,
        ...action.value,
        periodDates,
      };
    }
    case 'SET_USAGE_DATA': {
      const { periodicUsage } = mapResponseToState(action.value);
      const totalUsage = sum(periodicUsage.org.usage);
      return { ...state, periodicUsage, totalUsage };
    }
    case 'SET_ASSET_BANDWIDTH_DATA': {
      const assetBandwidthData = {
        usage: get(action.value, ['usage']),
        limit: get(action.value, ['limits', 'included']),
        uom: get(action.value, ['unitOfMeasure']),
      };
      return { ...state, assetBandwidthData };
    }
    case 'CHANGE_PERIOD': {
      return {
        ...state,
        selectedPeriodIndex: action.value,
        periodDates: periodToDates(state.periods[action.value]),
      };
    }
    case 'SET_LOADING': {
      return { ...state, isLoading: action.value };
    }
    case 'SWITCH_MAIN_TAB': {
      return {
        ...state,
        selectedMainTab: action.value,
        isAssetBandwidthTab: action.value === 'assetBandwidth',
      };
    }
    case 'SWITCH_SPACES_TAB': {
      return { ...state, selectedSpacesTab: action.value };
    }
    case 'SET_ERROR': {
      return { ...state, error: action.value };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

export const UsageProvider = ({ children, orgId }) => {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  return (
    <UsageStateContext.Provider value={{ ...state, orgId }}>
      <UsageDispatchContext.Provider value={dispatch}>{children}</UsageDispatchContext.Provider>
    </UsageStateContext.Provider>
  );
};

UsageProvider.propTypes = {
  orgId: PropTypes.string,
};

export const useUsageState = () => {
  const context = React.useContext(UsageStateContext);
  if (!context) {
    throw new Error('useUsageState must be used within a UsageProvider');
  }
  return context;
};

export const useUsageDispatch = () => {
  const context = React.useContext(UsageDispatchContext);
  if (!context) {
    throw new Error('useUsageDispatch must be used within a UsageProvider');
  }
  return context;
};
