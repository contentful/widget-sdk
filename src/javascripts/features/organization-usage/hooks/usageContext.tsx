import React, { createContext, useReducer, Dispatch, Context, ReactNode } from 'react';
import { mapResponseToState } from '../services/UsageService';
import { periodToDates, PeriodToDatesArgs, PeriodToDatesResponse } from '../utils/periodToDates';
import { sum, get } from 'lodash';
import { noop } from 'lodash/fp';

export const colours = ['#2E75D4', '#0EB87F', '#EA9005', '#8C53C2', '#CC3C52'];

type UsageContextStateType = {
  isLoading: boolean;
  error?: Error | null | undefined;
  isAssetBandwidthTab: boolean;
  selectedMainTab: string;
  selectedPeriodIndex: number;
  selectedSpacesTab: string;
  isTeamOrEnterpriseCustomer?: boolean;
  periodDates?: PeriodToDatesResponse;
  periodicUsage?: {
    org: {
      usage: number[];
    };
    apis: any;
  };
  totalUsage?: number;
  apiRequestIncludedLimit?: number;
  assetBandwidthData?: {
    usage: any;
    limit: any;
    uom: any;
  };
  spaceNames: Record<string, string>;
  spaceTypeLookup: Record<string, any>;
  periods: PeriodToDatesArgs[];
  hasSpaces?: boolean;
  orgId?: string;
};

export const initialState = {
  isLoading: true,
  error: null,
  isAssetBandwidthTab: false,
  selectedMainTab: 'apiRequest',
  selectedPeriodIndex: 0,
  selectedSpacesTab: 'cma',
  spaceNames: {},
  spaceTypeLookup: {},
  periods: [] as PeriodToDatesArgs[],
};

export const UsageStateContext = createContext(initialState) as Context<UsageContextStateType>;
export const UsageDispatchContext = createContext(noop) as Context<Dispatch<any>>;

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_CUSTOMER_TYPE': {
      return { ...state, isTeamOrEnterpriseCustomer: action.value };
    }
    case 'SET_ORG_DATA': {
      const { periods } = action.value;
      const periodDates = periodToDates(periods[state.selectedPeriodIndex]);
      return {
        ...state,
        ...action.value,
        periodDates,
      };
    }
    case 'SET_USAGE_DATA': {
      const { periodicUsage } = mapResponseToState(action.value);
      const totalUsage = sum(periodicUsage.org.usage);
      return { ...state, periodicUsage, totalUsage };
    }
    case 'SET_ASSET_BANDWIDTH_DATA': {
      const assetBandwidthData = {
        usage: get(action.value, ['usage']),
        limit: get(action.value, ['limits', 'included']),
        uom: get(action.value, ['unitOfMeasure']),
      };
      return { ...state, assetBandwidthData };
    }
    case 'CHANGE_PERIOD': {
      return {
        ...state,
        selectedPeriodIndex: action.value,
        periodDates: periodToDates(state.periods[action.value]),
      };
    }
    case 'SET_LOADING': {
      return { ...state, isLoading: action.value };
    }
    case 'SWITCH_MAIN_TAB': {
      return {
        ...state,
        selectedMainTab: action.value,
        isAssetBandwidthTab: action.value === 'assetBandwidth',
      };
    }
    case 'SWITCH_SPACES_TAB': {
      return { ...state, selectedSpacesTab: action.value };
    }
    case 'SET_ERROR': {
      return { ...state, error: action.value };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

type UsageProviderProps = {
  children: ReactNode;
  orgId: string;
};

export const UsageProvider = ({ children, orgId }: UsageProviderProps) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <UsageStateContext.Provider value={{ ...state, orgId }}>
      <UsageDispatchContext.Provider value={dispatch}>{children}</UsageDispatchContext.Provider>
    </UsageStateContext.Provider>
  );
};

export const useUsageState = () => {
  const context = React.useContext(UsageStateContext);
  if (!context) {
    throw new Error('useUsageState must be used within a UsageProvider');
  }
  return context;
};

export const useUsageDispatch = () => {
  const context = React.useContext(UsageDispatchContext);
  if (!context) {
    throw new Error('useUsageDispatch must be used within a UsageProvider');
  }
  return context;
};
