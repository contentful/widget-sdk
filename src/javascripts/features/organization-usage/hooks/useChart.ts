import { useEffect, useState, useRef } from 'react';

import * as echarts from 'echarts';

export const useChart = (props) => {
  const chartRef = useRef<HTMLDivElement | null>(null);
  const [chart, setChart] = useState(null);

  useEffect(() => {
    if (chartRef.current) {
      setChart(echarts.init(chartRef.current));
    }
  }, []);

  useEffect(() => {
    if (chart) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      chart.setOption(props);
    }
  });

  return chartRef;
};
