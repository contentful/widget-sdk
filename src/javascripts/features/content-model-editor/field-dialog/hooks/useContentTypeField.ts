import { useCallback, useMemo, useState } from 'react';
import { getWidgetSettings } from '../utils/helpers';
import { useForm } from 'core/hooks';
import {
  getInitialValueFormFields,
  getNodeValidationsFormFields,
  getSettingsFormFields,
  getValidationsFormFields,
} from '../utils/formFieldDefinitions';
import { ContentType, ContentTypeField, EditorInterfaceControl } from 'core/typings';

type ContentTypeFieldProps = {
  contentType: ContentType;
  field: ContentTypeField;
  submitFn: Function;
  widget: EditorInterfaceControl;
};

function useWidgetSettings(widget) {
  const [data, setData] = useState(() => getWidgetSettings(widget));
  const [pristine, setPristine] = useState(true);
  const setActualData = useCallback((newData) => {
    setData(newData);
    setPristine(false);
  }, []);

  return {
    pristine,
    data,
    setData: setActualData,
  };
}

export function useContentTypeField({
  submitFn,
  contentType,
  field,
  widget,
}: ContentTypeFieldProps) {
  const widgetSettings = useWidgetSettings(widget);
  const form = useForm({
    fields: {
      ...getSettingsFormFields(field, contentType),
      ...getValidationsFormFields(field),
      ...getNodeValidationsFormFields(field),
      ...getInitialValueFormFields(field),
    },
    submitFn,
  });

  return useMemo(
    () => ({
      contentType: form,
      widgetSettings: {
        data: widgetSettings.data,
        setData: widgetSettings.setData,
      },
      get isInvalid() {
        return form.form.invalid;
      },
      get isPristine() {
        return form.form.pristine && widgetSettings.pristine;
      },
    }),
    [form, widgetSettings]
  );
}
