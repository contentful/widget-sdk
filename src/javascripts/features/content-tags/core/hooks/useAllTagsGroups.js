import { useEffect, useState } from 'react';
import { groupByName, tagsPayloadToValues } from 'features/content-tags/editor/utils';
import { useReadTags } from 'features/content-tags/core/hooks/useReadTags';

const useAllTagsGroups = () => {
  const { data } = useReadTags();
  const [tagGroups, setTagGroups] = useState([]);
  useEffect(() => {
    const allDataTags = tagsPayloadToValues(data);
    setTagGroups(Object.keys(groupByName(allDataTags)));
  }, [data, setTagGroups]);
  return tagGroups;
};

export { useAllTagsGroups };
