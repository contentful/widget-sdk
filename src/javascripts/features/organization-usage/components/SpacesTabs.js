import React from 'react';
import { Tabs, Tab, TabPanel } from '@contentful/forma-36-react-components';
import tokens from '@contentful/forma-36-tokens';
import { Grid, GridItem } from '@contentful/forma-36-react-components/dist/alpha';
import { css } from 'emotion';
import { SpacesTable } from './SpacesTable';
import { SpacesBarChart } from './SpacesBarChart';

import { track } from 'analytics/Analytics';
import { useUsageState, useUsageDispatch } from '../hooks/usageContext';

const styles = {
  tabPanel: css({
    paddingTop: tokens.spacing2Xl,
    paddingBottom: tokens.spacing2Xl,
  }),
};

const tabsData = [
  {
    id: 'cma',
    title: 'CMA Requests',
  },
  {
    id: 'cda',
    title: 'CDA Requests',
  },
  {
    id: 'cpa',
    title: 'CPA Requests',
  },
  {
    id: 'gql',
    title: 'GraphQL Requests',
  },
];

export const SpacesTabs = () => {
  const { selectedSpacesTab } = useUsageState();
  const dispatch = useUsageDispatch();

  const handleSelected = (id) => {
    track('usage:space_tab_selected', { old: selectedSpacesTab, new: id });
    dispatch({ type: 'SWITCH_SPACES_TAB', value: id });
  };

  return (
    <>
      <Tabs withDivider={true}>
        {tabsData.map((item) => (
          <Tab
            id={item.id}
            key={item.id}
            selected={selectedSpacesTab === item.id}
            onSelect={handleSelected}>
            {item.title}
          </Tab>
        ))}
      </Tabs>
      <TabPanel id="api-usage-tab-panel" className={styles.tabPanel}>
        <Grid columns={'1fr 2fr'}>
          <GridItem>
            <SpacesTable />
          </GridItem>
          <GridItem>
            <SpacesBarChart />
          </GridItem>
        </Grid>
      </TabPanel>
    </>
  );
};
