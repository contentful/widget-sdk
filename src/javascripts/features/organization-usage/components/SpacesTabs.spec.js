import React from 'react';
import { render, fireEvent, within } from '@testing-library/react';
import * as echarts from 'echarts';

import { track } from 'analytics/Analytics';
import { SpacesTabs } from './SpacesTabs';

import { UsageStateContext, UsageDispatchContext } from '../hooks/usageContext';
import { sum } from 'lodash';

jest.mock('echarts', () => ({
  init: jest.fn(),
}));

const ORG_USAGE = [147, 651];
const CMA_USAGE = [48, 46];
const CMA_USAGE_DELETED_SPACE = [48, 56];
const CDA_USAGE = [48, 66];

const defaultData = {
  spaceNames: { cmaSpace: 'CMASpace', cdaSpace: 'CDASpace' },
  periodDates: ['1 Feb', '2 Feb'],
  periodicUsage: {
    org: { usage: ORG_USAGE },
    apis: {
      cma: {
        items: [
          {
            usage: CMA_USAGE,
            sys: {
              space: {
                sys: {
                  id: 'cmaSpace',
                },
              },
            },
          },
          {
            usage: CMA_USAGE_DELETED_SPACE,
            sys: {
              space: {
                sys: {
                  id: 'Deleted space',
                },
              },
            },
          },
        ],
      },
      cda: {
        items: [
          {
            usage: CDA_USAGE,
            sys: {
              space: {
                sys: {
                  id: 'cdaSpace',
                },
              },
            },
          },
        ],
      },
    },
  },
  isPoC: {
    cmaSpace: true,
    cdaSpace: false,
  },
  selectedSpacesTab: 'cma',
  totalUsage: sum(ORG_USAGE),
};

const renderComp = (data, dispatch) => {
  return render(
    <MockPovider {...data} dispatch={dispatch}>
      <SpacesTabs />
    </MockPovider>
  );
};

const MockPovider = (data) => {
  const { dispatch, children } = data;
  return (
    <UsageStateContext.Provider value={data}>
      <UsageDispatchContext.Provider value={dispatch}>{children}</UsageDispatchContext.Provider>
    </UsageStateContext.Provider>
  );
};

MockPovider.defaultProps = {
  dispatch: () => {},
};

describe('SpacesTabs', () => {
  it('renders correctly', () => {
    const container = renderComp(defaultData);
    expect(container).toMatchSnapshot();
  });

  describe('by default', () => {
    it('renders CMA data correctly', () => {
      const { getByTestId, getByText } = renderComp(defaultData);

      const table = getByTestId('api-usage-table');
      expect(table).toBeInTheDocument();

      const usage = `${sum(CMA_USAGE)}`;
      expect(getByText(usage)).toBeInTheDocument();

      const usageDeletedSpace = `${sum(CMA_USAGE_DELETED_SPACE)}`;
      expect(getByText(usageDeletedSpace)).toBeInTheDocument();

      expect(getByText('CMASpace')).toBeInTheDocument();
    });

    describe('SpacesTable', () => {
      it('shows deleted space', () => {
        const { getByText } = renderComp(defaultData);
        expect(getByText('Deleted space')).toBeInTheDocument();
      });

      it('shows one space as PoC', () => {
        const { getAllByText } = renderComp(defaultData);
        const tags = getAllByText('POC');
        expect(tags).toHaveLength(1);
        fireEvent.mouseOver(tags[0]);
        expect(getAllByText('Proof of concept')).toHaveLength(1);
      });
    });

    it('renders a bar chart', () => {
      const { getByTestId } = renderComp(defaultData);
      expect(getByTestId('api-usage-bar-chart')).toBeInTheDocument();

      expect(echarts.init).toHaveBeenCalledTimes(1);
    });
  });

  describe('on CDA tab click', () => {
    it('switch to the correct tab', () => {
      const dispatchSpy = jest.fn();
      const { getByText } = renderComp(defaultData, dispatchSpy);

      expect(dispatchSpy).not.toHaveBeenCalled();
      fireEvent.click(getByText('CDA Requests'));
      expect(dispatchSpy).toHaveBeenCalledWith({ type: 'SWITCH_SPACES_TAB', value: 'cda' });
    });

    it('renders CDA data correctly', () => {
      const updatedData = {
        ...defaultData,
        selectedSpacesTab: 'cda',
      };

      const { getByText, getByTestId } = renderComp(updatedData);

      const table = getByTestId('api-usage-table');
      expect(table).toBeInTheDocument();

      const usage = `${sum(CDA_USAGE)}`;
      expect(getByText(usage)).toBeInTheDocument();

      expect(getByText('CDASpace')).toBeInTheDocument();
    });

    it('does not show any PoC spaces', () => {
      const updatedData = {
        ...defaultData,
        selectedSpacesTab: 'cda',
      };

      const { queryByText } = renderComp(updatedData);
      expect(queryByText('POC')).toBeNull();
    });

    it('renders a bar chart and initializes once', () => {
      const { getByTestId, getByText } = renderComp(defaultData);
      expect(echarts.init).toHaveBeenCalledTimes(1);

      fireEvent.click(getByText('CDA Requests'));

      expect(getByTestId('api-usage-bar-chart')).toBeInTheDocument();
      expect(echarts.init).toHaveBeenCalledTimes(1);
    });

    it('sends a cma to cda tracking event', () => {
      const { getByText } = renderComp(defaultData);

      fireEvent.click(getByText('CDA Requests'));

      expect(track).toHaveBeenCalledTimes(1);
      expect(track).toHaveBeenCalledWith('usage:space_tab_selected', { new: 'cda', old: 'cma' });
    });
  });

  describe('with empty data', () => {
    const data = {
      spaceNames: {},
      period: [],
      periodicUsage: {
        apis: {
          cma: { items: [] },
          cda: { items: [] },
        },
        org: { usage: [] },
      },
      isPoC: {},
      selectedSpacesTab: 'cma',
    };

    it('renders an empty table', () => {
      const { getByTestId, queryByTestId } = renderComp(data);
      const table = getByTestId('api-usage-table');
      expect(table).toBeInTheDocument();
      expect(table.firstElementChild).toBeInTheDocument();
      within(table.firstChild).getByText('Current usage period');
      expect(queryByTestId('api-usage-table-row')).toBeNull();
    });

    it('renders a chart', () => {
      const { getByTestId } = renderComp(data);
      expect(getByTestId('api-usage-bar-chart')).toBeInTheDocument();
      expect(echarts.init).toHaveBeenCalledTimes(1);
    });

    it('can switch tab', () => {
      const { getByTestId, getByText } = renderComp(data);
      fireEvent.click(getByText('CDA Requests'));
      const table = getByTestId('api-usage-table');
      expect(table).toBeInTheDocument();

      expect(getByTestId('api-usage-bar-chart')).toBeInTheDocument();
    });
  });

  describe('loading state', () => {
    it('shows skeleton rows', () => {
      const { getAllByTestId } = renderComp({ ...defaultData, isLoading: true });

      getAllByTestId('cf-ui-skeleton-form').forEach((ele) => {
        expect(ele).toBeVisible();
      });
    });
  });
});
