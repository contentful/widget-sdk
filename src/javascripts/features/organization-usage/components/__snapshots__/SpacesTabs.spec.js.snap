// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpacesTabs renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  height: 512px;
  width: 100%;
}

.emotion-1 {
  padding-top: 3rem;
  padding-bottom: 3rem;
}

<body>
    <div>
      <div
        class="Tabs__Tabs___3Cp8m Tabs__Tabs__with-divider___2Mx3Z"
        data-test-id="cf-ui-tabs"
        role="tablist"
      >
        <div
          aria-controls="cma"
          aria-selected="true"
          class="Tabs__Tab___1SiYI Tabs__Tab__selected___3erbm"
          data-test-id="cf-ui-tab"
          role="tab"
          tabindex="0"
        >
          CMA Requests
        </div>
        <div
          aria-controls="cda"
          aria-selected="false"
          class="Tabs__Tab___1SiYI"
          data-test-id="cf-ui-tab"
          role="tab"
          tabindex="0"
        >
          CDA Requests
        </div>
        <div
          aria-controls="cpa"
          aria-selected="false"
          class="Tabs__Tab___1SiYI"
          data-test-id="cf-ui-tab"
          role="tab"
          tabindex="0"
        >
          CPA Requests
        </div>
        <div
          aria-controls="gql"
          aria-selected="false"
          class="Tabs__Tab___1SiYI"
          data-test-id="cf-ui-tab"
          role="tab"
          tabindex="0"
        >
          GraphQL Requests
        </div>
      </div>
      <div
        class="emotion-1"
        data-test-id="cf-ui-tab-panel"
        id="api-usage-tab-panel"
        role="tabpanel"
      >
        <div
          class="Grid__Grid___2_rpl"
          data-test-id="cf-ui-grid"
          style="grid-template-columns: 1fr 2fr; grid-template-rows: auto;"
        >
          <div>
            <table
              cellpadding="0"
              cellspacing="0"
              class="Table__Table___3vKIR"
              data-test-id="api-usage-table"
            >
              <thead
                class=""
                data-test-id="cf-ui-table-head"
              >
                <tr
                  class="TableRow__TableRow___3kcNC"
                  data-test-id="cf-ui-table-row"
                >
                  <th
                    align="left"
                    class="TableCell__TableCell___Wou8a TableCell__TableCell--head___1HhvZ"
                    data-test-id="cf-ui-table-cell"
                    style="top: 0px;"
                  >
                    Space
                  </th>
                  <th
                    align="left"
                    class="TableCell__TableCell___Wou8a TableCell__TableCell--head___1HhvZ"
                    data-test-id="cf-ui-table-cell"
                    style="top: 0px;"
                  >
                    Current usage period
                  </th>
                  <th
                    align="left"
                    class="TableCell__TableCell___Wou8a TableCell__TableCell--head___1HhvZ"
                    data-test-id="cf-ui-table-cell"
                    style="top: 0px;"
                  >
                    Total API requests (%)
                  </th>
                </tr>
              </thead>
              <tbody
                data-test-id="cf-ui-table-body"
              >
                <tr
                  class="TableRow__TableRow___3kcNC"
                  data-test-id="api-usage-table-row"
                >
                  <td
                    align="left"
                    class="TableCell__TableCell___Wou8a"
                    data-test-id="cf-ui-table-cell"
                    style="top: 0px;"
                  >
                    <div>
                      CMASpace
                    </div>
                    <div>
                      <span>
                        <div
                          class="Tag__Tag___Y-myd Tag__Tag--muted___1Uba5"
                          data-test-id="cf-ui-tag"
                        >
                          POC
                        </div>
                      </span>
                    </div>
                  </td>
                  <td
                    align="left"
                    class="TableCell__TableCell___Wou8a"
                    data-test-id="cf-ui-table-cell"
                    style="vertical-align: middle;"
                  >
                    94
                  </td>
                  <td
                    align="left"
                    class="TableCell__TableCell___Wou8a"
                    data-test-id="cf-ui-table-cell"
                    style="color: rgb(46, 117, 212); vertical-align: middle;"
                  >
                    12
                    %
                  </td>
                </tr>
                <tr
                  class="TableRow__TableRow___3kcNC"
                  data-test-id="api-usage-table-row"
                >
                  <td
                    align="left"
                    class="TableCell__TableCell___Wou8a"
                    data-test-id="cf-ui-table-cell"
                    style="top: 0px;"
                  >
                    <div>
                      Deleted space
                    </div>
                    <div />
                  </td>
                  <td
                    align="left"
                    class="TableCell__TableCell___Wou8a"
                    data-test-id="cf-ui-table-cell"
                    style="vertical-align: middle;"
                  >
                    104
                  </td>
                  <td
                    align="left"
                    class="TableCell__TableCell___Wou8a"
                    data-test-id="cf-ui-table-cell"
                    style="color: rgb(14, 184, 127); vertical-align: middle;"
                  >
                    13
                    %
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div>
            <div
              class="emotion-0"
              data-test-id="api-usage-bar-chart"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": .emotion-0 {
  height: 512px;
  width: 100%;
}

.emotion-1 {
  padding-top: 3rem;
  padding-bottom: 3rem;
}

<div>
    <div
      class="Tabs__Tabs___3Cp8m Tabs__Tabs__with-divider___2Mx3Z"
      data-test-id="cf-ui-tabs"
      role="tablist"
    >
      <div
        aria-controls="cma"
        aria-selected="true"
        class="Tabs__Tab___1SiYI Tabs__Tab__selected___3erbm"
        data-test-id="cf-ui-tab"
        role="tab"
        tabindex="0"
      >
        CMA Requests
      </div>
      <div
        aria-controls="cda"
        aria-selected="false"
        class="Tabs__Tab___1SiYI"
        data-test-id="cf-ui-tab"
        role="tab"
        tabindex="0"
      >
        CDA Requests
      </div>
      <div
        aria-controls="cpa"
        aria-selected="false"
        class="Tabs__Tab___1SiYI"
        data-test-id="cf-ui-tab"
        role="tab"
        tabindex="0"
      >
        CPA Requests
      </div>
      <div
        aria-controls="gql"
        aria-selected="false"
        class="Tabs__Tab___1SiYI"
        data-test-id="cf-ui-tab"
        role="tab"
        tabindex="0"
      >
        GraphQL Requests
      </div>
    </div>
    <div
      class="emotion-1"
      data-test-id="cf-ui-tab-panel"
      id="api-usage-tab-panel"
      role="tabpanel"
    >
      <div
        class="Grid__Grid___2_rpl"
        data-test-id="cf-ui-grid"
        style="grid-template-columns: 1fr 2fr; grid-template-rows: auto;"
      >
        <div>
          <table
            cellpadding="0"
            cellspacing="0"
            class="Table__Table___3vKIR"
            data-test-id="api-usage-table"
          >
            <thead
              class=""
              data-test-id="cf-ui-table-head"
            >
              <tr
                class="TableRow__TableRow___3kcNC"
                data-test-id="cf-ui-table-row"
              >
                <th
                  align="left"
                  class="TableCell__TableCell___Wou8a TableCell__TableCell--head___1HhvZ"
                  data-test-id="cf-ui-table-cell"
                  style="top: 0px;"
                >
                  Space
                </th>
                <th
                  align="left"
                  class="TableCell__TableCell___Wou8a TableCell__TableCell--head___1HhvZ"
                  data-test-id="cf-ui-table-cell"
                  style="top: 0px;"
                >
                  Current usage period
                </th>
                <th
                  align="left"
                  class="TableCell__TableCell___Wou8a TableCell__TableCell--head___1HhvZ"
                  data-test-id="cf-ui-table-cell"
                  style="top: 0px;"
                >
                  Total API requests (%)
                </th>
              </tr>
            </thead>
            <tbody
              data-test-id="cf-ui-table-body"
            >
              <tr
                class="TableRow__TableRow___3kcNC"
                data-test-id="api-usage-table-row"
              >
                <td
                  align="left"
                  class="TableCell__TableCell___Wou8a"
                  data-test-id="cf-ui-table-cell"
                  style="top: 0px;"
                >
                  <div>
                    CMASpace
                  </div>
                  <div>
                    <span>
                      <div
                        class="Tag__Tag___Y-myd Tag__Tag--muted___1Uba5"
                        data-test-id="cf-ui-tag"
                      >
                        POC
                      </div>
                    </span>
                  </div>
                </td>
                <td
                  align="left"
                  class="TableCell__TableCell___Wou8a"
                  data-test-id="cf-ui-table-cell"
                  style="vertical-align: middle;"
                >
                  94
                </td>
                <td
                  align="left"
                  class="TableCell__TableCell___Wou8a"
                  data-test-id="cf-ui-table-cell"
                  style="color: rgb(46, 117, 212); vertical-align: middle;"
                >
                  12
                  %
                </td>
              </tr>
              <tr
                class="TableRow__TableRow___3kcNC"
                data-test-id="api-usage-table-row"
              >
                <td
                  align="left"
                  class="TableCell__TableCell___Wou8a"
                  data-test-id="cf-ui-table-cell"
                  style="top: 0px;"
                >
                  <div>
                    Deleted space
                  </div>
                  <div />
                </td>
                <td
                  align="left"
                  class="TableCell__TableCell___Wou8a"
                  data-test-id="cf-ui-table-cell"
                  style="vertical-align: middle;"
                >
                  104
                </td>
                <td
                  align="left"
                  class="TableCell__TableCell___Wou8a"
                  data-test-id="cf-ui-table-cell"
                  style="color: rgb(14, 184, 127); vertical-align: middle;"
                >
                  13
                  %
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div>
          <div
            class="emotion-0"
            data-test-id="api-usage-bar-chart"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
